### DISCUSSION TITLE
Cloudflare e sottodominio

### DISCUSSION THREAD

**Discussion by gasparirob** [indomus/forum#3094] [HUB: Home Assistant]
Ciao a tutti. 
Ho seguito la guida per la migrazione da duckdns a cloudflare.

È tutto perfettamente funzionante già da diversi mesi e senza più i continui down di duckdns.

Adesso chiedo: è possibile creare un sottodominio in modo da veicolare il traffico verso un altro servizio attivo all'interno della mia rete? 

Nello specifico (ma il discorso potrebbe essere più generale), a partire dal mio dominio miodominio.com (che uso per accedere ad Home Assistant) vorrei creare il sottodominio jelly.miodominio.com per accedere a Jellyfin dall'esterno (sfruttando il tunnel di cloudflare) 

PS: Jellyfin, nel mio caso, è un addon di Home Assistant, qualora questo possa fare qualche differenza. 

Grazie mille a tutti

**Comment by snakuzzo** [comment]
Leggi la documentazione dell'addon... è scritto tutto lì.

https://github.com/brenner-tobias/addon-cloudflared/blob/main/cloudflared/DOCS.md#overview-add-on-configuration

**Reply by gasparirob** [reply]
In realtà l'avevo letta... ma mi aspettavo qualche configurazione da effettuare lato cloudflare.
Invece è bastato aggiungere l'additional host al file di configurazione... 

Grazie mille

**Reply by lux90ita** [reply]
puoi farla anche lato cloudflare
andando su    Zero Trust \ Network \ Tunnels \ Public Hostname Page
andando qui vedi quelle che hai creato (Attento che l'addon non le elimina quindi se hai fatto dei test devi rimuoverle a mano)
un consiglio è creare anche delle regole WAF per Country

**Comment by vincenzocoletta** [comment]
Ciao, mi aggancio a questo thread perché la tematica è la stessa ma vorrei capire se esiste una soluzione diversa: una volta migrato tutto su Cloudflare (è tutto funziona in maniera egregia, grazie e complimenti ancora per le guide!) posso continuare ad utilizzare un diverso DDNS (ad esempio DynDNS) aggiornando l'indirizzo dinamico tramite il mio router (Vodafone Station) per accedere da remoto ad altri PC/dispositivi della mia rete? Non ci avevo ancora provato a distanza di mesi dalla migrazione su Cloudflare ed ora mi sembra che questa opzione non sia più percorribile... qualcuno può confermare o smentire prima che continui a sbatterci su la testa?
A presto e grazie,
V!

**Reply by snakuzzo** [reply]
Certo che puoi continuare ad usarlo.

**Reply by vincenzocoletta** [reply]
Ti ringrazio, eppure ti assicuro che non funziona... come se il tunneling di Cloudflare escludesse la possibilità di altri indirizzamenti DNS. L'associazione all'IP è corretta ed aggiornata (verificata su DynDNS) le porte sono correttamente associate su router ma non mi lascia entrare. Ma mi fido e proverò a sbatterci ancora un pò la testa...

**Reply by snakuzzo** [reply]
Credimi...funziona. Perché io uso sia Cloudflare che Dynamic DNS con port forwarding...e funziona tutto perfettamente :)

**Reply by vincenzocoletta** [reply]
Allora devo riuscirci per forza, non mollo!
Ne approfitto però per chiederti un altra cosa sempre in tema: qualora volessi utilizzare il reindirizzamento di Cloudflare come da tuoi preziosi suggerimenti (vedi su), posso agganciare allo stesso sottodominio più macchine nella stessa rete? Un po come si fa con l'associazione porte di DynDNS quindi qualcosa del tipo casa.dyndns.com:xxx e casa.dyndns.com:yyy che puntano a macchine diverse. Ovviamente non necessariamente con le porte, il mio è solo un esempio ma l'esigenza sarebbe quella di tenere lo stesso sottodominio per macchine con diverso ip locale...
Ancora grazie 1000!,
V!

**Reply by snakuzzo** [reply]
Sì, puoi farlo, ma non con le porte, come hai scritto...ma con CNAME diversi.
Puoi usare lo stesso tunnel (quindi in sostanza lo stesso container, daemon o quello che è) con una configurazione con più CNAME)...tipo così (da notare come ho commentato la configurazione standard e messo le N entry per i vari CNAME)...

```yaml
tunnel: xxxxx-xxx-xxx-xxx-xxxxxxxxxx
credentials-file: /home/nonroot/.cloudflared/xxxxx-xxx-xxx-xxx-xxxxxxxxxx.json
#url: http://192.168.1.10:8124

ingress:
  - hostname: homoeassistant.tuodominio.com
    service: http://192.168.1.10:8124
  - hostname: plex.tuodominio.com
    service: http://192.168.1.10:11080
  - service: http_status:404
```

In questo modo è facile capire come accederai dall'esterno ai vari servizi ;)

**Reply by vincenzocoletta** [reply]
Okok, grazie. Se ho capito bene così però avrò sottodomini diversi per servizi diversi mentre o me servirebbe avere lo stesso sottodominio, per esempio telecamere.miodominio.com, per accedere alla cam1, cam2, cam3,... Non esiste un modo che consenta di fare qualcosa del tipo https://telecamere.miodominio.it/camx (con x=1, 2 e 3)?

**Reply by snakuzzo** [reply]
Questo non c'entra nulla con il tunnel...lì è una questione da gestire a livello di redirect. E comunque, anche se riuscissi a farlo, potresti avere dei comportamenti non voluti.
Se l'obiettivo è quello di accedere alle cam, premesso che io non le esporrei MAI NELLA VITA così come hai detto tu (**è da pazzi**), non avresti alcun problema usando CNAME diversi

**Reply by vincenzocoletta** [reply]
Accetto il suggerimento ma... tu come come le esporresti? In cloud ancora peggio e poi non potrei disporre dei settings che sono accessibili sono con accesso diretto e non tramite app...

**Reply by snakuzzo** [reply]
1. Non prenderei cam che si appoggiano ad un servizio in Cloud. Es. Ho preso delle YI su cui ho firmware alternativo e con il quale ho disabilitato totalmente il Cloud.

2. Non le esporrei direttamente. Le raggiungerei tramite un server VPN tirato su dove ho le cam, oppure le esporrei tramite un altro servizio con auth gestita come si deve (es. tramite Home Assistant).

Ora però, discorso chiuso perché siamo totalmente off topic.

Ciao.

**Reply by vincenzocoletta** [reply]
Ops! Hai ragione, scusami… comunque ottimi suggerimenti, grazie! Passo e chiudo…

**Comment by omarantelmi** [comment]
Buongiorno
Mi collego alla questione perchè mi sembra che sia inerente

Ho configurato con la vostra guida (fantastica) cloudflare per accedere ad HA con https://miodominio.xx, questo punta al mio server dedicato ad HA
Ho un altro server, con indirizzo ip locale differente ovviamente, che vorrei raggiungere sempre con lo stesso indirizzo ma a questo punto preferirei riorganizzare il tutto e avere sull'indirizzo principale (https://miodominio.xx) accesso ad HOMEPAGE servizio che ho su 192.168.1.114:xxxx + IMMICH che è su 192.168.1.114:yyyy + eventuali altri servizi che istanzierò sempre sulla stessa macchina 192.168.1.114:xxxx + HA che va su 192.168.1.81:xxxx

A questo punto vorrei che il tutto avesse accesso come indirizzi i seguenti: 
https://homepage.miodominio.xx
https://immich.miodominio.xx
https://altriservizi.miodominio.xx
https://ha.miodominio.xx

O in alternativa
https://miodominio.xx/homepage
https://miodominio.xx/immich
https://miodominio.xx/altroservizio
https://miodominio.xx/ha

In base a quella che potrebbe essere la via corretta

Purtroppo non ho trovato un granchè per arrivare a questa soluzione, ovviamente colpa anche della mia poca conoscenza di tutto ciò, vi sarei grato se poteste indirizzarmi in modo corretto

**Reply by snakuzzo** [reply]
La prima che hai detto.
La seconda è un casino e non è detto che funzioni, perché dipende anche da com'è scritta l'applicazione che andrai a proxare.

Leggi la doc...

https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/configure-tunnels/local-management/configuration-file/#file-structure-for-public-hostnames

**Reply by omarantelmi** [reply]
Era molto più semplice di quello che pensassi ma ad ogni modo la doc che mi hai indicato mi ha aiutato, scrivo per chiunque cercasse una soluzione simile

Dopo aver eseguito la guida indomus basterà configurare all'interno di homeassistant, nel componente aggiuntivo cloudflare, la parte configurazione
Nello specifico
1) ho fermato il servizio e ho modificato external_hostname da miodominio.xx a ha.miodominio.xx
2) Ho inserito gli altri servizi in additional_hosts esattamente così:
  - hostname: homepage.miodominio.xx
    service: http://iplocale:port
  - hostname: immich.miodominio.xx
    service: http://iplocale:port
3) ho avviato nuovamente il servizio cloudfare

Finito, tutto funziona così come volevo, sono giorni che ci sbatto ma in realtà è di una semplicità disarmante

**Reply by gasparirob** [reply]
Ciao, avevo già provato a fare la stessa cosa che hai fatto tu, ma nel momento in cui punto ad un servizio esterno a dove si trova HA, Cloudflare mi da errore in fase di avvio:
`Validation failed: http://192.168.1.181/ is an invalid address, ingress rules don't support proxying to a different path on the origin service. The path will be the same as the eyeball request's path`
(HA si trova su 192.168.1.172)

**Comment by lux90ita** [comment]
puoi farlo anche da interfaccia grafica andando su "zero trust" di cloudflare

**Reply by omarantelmi** [reply]
Potrei, tuttavia ho seguito la guida indomus la quale mi ha creato un tunnel locale, quindi se ho capito bene la risoluzione è da trovare nel tunnel già creato e sicuramente da lato locale
Ad ogni modo ho risolto come scritto sopra

**Reply by snakuzzo** [reply]
La risoluzione viene gestita da Cloudflare tramite i vari cname di tipo proxied. Poi passando dal tunnel aperto dal tuo "addon" (in questo caso) cloudflared veicola il traffico verso i vari ip:porta che hai definito.
Ti consiglio di aggiungere la gestione del 404... è indicato sempre nella doc che ti ho linkato oggi

**Reply by omarantelmi** [reply]
Si lo avevo letto più volte e alla fine l'ho dimenticato, grazie mille lo aggiungo subito

**Reply by lux90ita** [reply]
> Potrei, tuttavia ho seguito la guida indomus la quale mi ha creato un tunnel locale, quindi se ho capito bene la risoluzione è da trovare nel tunnel già creato e sicuramente da lato locale Ad ogni modo ho risolto come scritto sopra

in realtà è come ha detto snakuzzo (La risoluzione viene gestita da Cloudflare tramite i vari cname di tipo proxied. Poi passando dal tunnel aperto dal tuo "addon" (in questo caso) cloudflared veicola il traffico verso i vari ip:porta che hai definito.)

se vai a vedere nello zero trust all'interno del tunnel locale creato (Vedrai che ci sono anche i test che magari hai fatto, siccome di default non vengono cancellati) rischio esporre servizi non voluti :D
e per di più creare magari delle regole WAF in maniera semplificata
cosa che non viene scritta in nessuna guida ma esporre cose su internet senza un adeguata protezione non è un ottima idea

**Comment by federoco** [comment]
Ciao, anche a me funziona tutto seguendo la validissima [guida](https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-via-reti-provider-tradizionali-o-che-usino-nat-via-cloudflare/) proposta e docker compose.

Ho solo una perplessità in merito all'aggiornamento della App. 
Mi confermate che si può procede come sempre?
Ovvero con i comandi:
1) docker compose pull cloudflared
2) docker compose up -d cloudflared
3) docker image prune -f
O meglio aggiungere un "--force-recreate" al comando 2)?
Grazie.

**Reply by indomus** [reply]
corretta la lista di comandi, non c'è bisogno di force
